File: webapp/Locus.Core/Services/ProjectService.cs
using Locus.Core.DTO;
using Locus.Core.Helpers;
using Locus.Core.Models;
using Locus.Core.Repositories;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Web;

//Service for project operations.
namespace Locus.Core.Services
{
    //Interface which contains methods for each CRUD operation
    public interface IProjectService
    {
        
        Project Save(Project project, string user);

        Project Update(Project project, string user);

        Project Delete(int idProject, string user);

        List<Project> GetAll();

        Project Get(int idProject);

        List<Project> GetActives(string username,bool isAdmin);

        List<Project> GetInactives();

        Project Enable(int id, string user);

        void uploadFile();

        List<ChangeLog> ProjectChangeLogs(int id);

        ChangeLog Restore(Project project, int version, string user);

        List<ProjectDTO> GetProjectDTO();

        byte[] GenerateProjectReportExcel(int id, string date, int executionId);

        byte[] DownloadExecutionReport(int id, string date, int executionId);

        AssignedStatusDTO ExcelRequestor();

        void RestoreSteps();
    }
    //Class which implements IProjectService's methods and instance to IProjectRepository
    public class ProjectService : IProjectService
    {
        //Instance of IProjectRepository
        private readonly IProjectRepository _projectRepository;

        

        //Constructor of ProjectService and initialize projectRepository
        public ProjectService(IProjectRepository projectRepository)
        {
            _projectRepository = projectRepository;
        }

        //Delete method calls to Delete Method From projectRepository.
        public Project Delete(int idProject, string user)
        {
           return  _projectRepository.Delete(idProject, user);
        }

        public Project Enable(int id, string user)
        {
            return _projectRepository.Enable(id, user);
        }

        public byte[] GenerateProjectReportExcel(int id, string date, int executionId)
        {
            Project project = _projectRepository.Get(id);
            ExcelCreator excelCreator = new ExcelCreator();

            return excelCreator.GenerateExcel(project, date, executionId);
        }

        public byte[] DownloadExecutionReport(int id, string date, int executionId)
        {
            Project project = _projectRepository.Get(id);
            ExcelCreator excelCreator = new ExcelCreator();

            return excelCreator.GenerateExecutionReport(project, date, executionId);
        }

        //Get method calls to Get Method From projectRepository.
        public Project Get(int idProject)
        {
            return _projectRepository.Get(idProject);
        }

        public List<Project> GetActives(string username,bool isAdmin)
        {
            return _projectRepository.GetActives(username, isAdmin);
        }

        //GetAll method calls to GetAll Method From projectRepository.
        public List<Project> GetAll()
        {
            return _projectRepository.GetAll();
        }

        public List<Project> GetInactives()
        {
            return _projectRepository.GetInactives();
        }

        public List<ProjectDTO> GetProjectDTO()
        {
            return _projectRepository.GetProjectDTO();
        }

        public List<ChangeLog> ProjectChangeLogs(int id)
        {
            return _projectRepository.ProjectChangeLogs(id);
        }

        public ChangeLog Restore(Project project, int version, string user)
        {
            return _projectRepository.Restore(project, version, user);
        }

        //Save method calls to Save Method From projectRepository.
        public Project Save(Project project, string user)
        {
            return _projectRepository.Save(project, user);
        }
        //Update method calls to Update Method From projectRepository.
        public Project Update(Project project, string user)
        {
            return _projectRepository.Update(project, user);
        }

        public void uploadFile()
        {

            var httpRequest = HttpContext.Current.Request;
            string filePathImage = "";
            foreach(string file in httpRequest.Files)
            {
                var postedFile = httpRequest.Files[file];
                var path = System.Web.HttpContext.Current.Server.MapPath("~\\Content\\img\\");
                filePathImage = file+".png";

                var filePath = path + filePathImage;
                postedFile.SaveAs(filePath);
            }
        }

        public AssignedStatusDTO ExcelRequestor()
        {
            return Helpers.ExcelRequestor.Requestor();
        }

        public void RestoreSteps()
        {
            _projectRepository.RestoreSteps();
        }
    }
}

File: webapp/Locust/Scripts/app/services/ProjectService.js
(function (angular, undefined) {
    'use strict';

    // Get services module
    var app = angular.module(appName);

    app.service('ProjectService', ['$resource', function ($resource) {
        
        var noop = angular.noop;
        var resource = $resource('/api/Project/:action/:id',
            { id: '@Id' }, {
                Save: {
                    method: 'POST'
                },
                GetActives: {
                    method: 'GET',
                    isArray: true
                },
                GetInactives: {
                    method: 'GET',
                    isArray: true
                },
                Update: {
                    method: 'PUT'
                },
                Upload: {
                    method: 'POST'
                },
                Delete: {
                    method: 'DELETE',
                    params: {
                        id: "@id"
                    }
                },
                Enable: {
                    method: 'GET',
                    isArray: false
                },
                Get: {
                    method: 'GET',
                    isArray: false
                },
                GetAll: {
                    method: 'GET',
                    isArray: true
                },
                uploadFile: {
                    method: 'POST'
                },
                ProjectChangeLogs: {
                    method: 'GET',
                    isArray: true
                },
                Restore: {
                    method: 'POST'
                },
                GetDashboard: {
                    method: 'GET',
                    isArray: false
                },
                GetForUsersList: {
                    method: 'GET',
                    isArray: false
                },
                RequestExcel: {
                    method: 'GET',
                    timeout: 5000,
                    isArray: false
                }
            });
            self = this;

        self.GetAll = function (success, error) {
            return resource.GetAll({ action: 'GetAll' }, success || noop, error || noop);
        };
        self.ProjectChangeLogs = function (id, success, error) {
            return resource.ProjectChangeLogs({ action: 'ProjectChangeLogs', id: id }, success || noop, error || noop);
        };
        self.GetActives = function (success, error) {
            return resource.GetActives({ action: 'GetActives' }, success || noop, error || noop);
        };
        self.GetForUsersList = function (success, error) {
            return resource.GetActives({ action: 'GetForUsersList' }, success || noop, error || noop);
        };

        self.Restore = function (entity, version, success, error) {
            return resource.Restore({ action: 'Restore', version: version }, entity, success || noop, error || noop);
        };

        self.GetInactives = function (success, error) {
            return resource.GetInactives({ action: 'GetInactives' }, success || noop, error || noop);
        };
        self.Get = function (id, success, error) {
            return resource.Get({ action: 'Get', id: id }, success || noop, error || noop);
        };
        self.Enable = function (id, success, error) {
            return resource.Enable({ action: 'Enable', id: id }, success || noop, error || noop);
        };
        self.Delete = function (id, success, error) {
            return resource.Delete({ action: 'Delete', id: id }, success || noop, error || noop);
        };

        self.Save = function (entity, success, error) {
            return resource.Save({ action: 'Save' }, entity, success || noop, error || noop);
        };

        self.pushFile = function (file, success, error) {
            var formdata = new FormData();
            formdata.append(file.name, file);
            return resource.uploadFile({ action: 'uploadFile' }, formdata, success || noop, error || noop);
        }

        self.Update = function (entity, success, error) {
            return resource.Update({ action: 'Update' }, entity, success || noop, error || noop);
        };

        self.GetDashboard = function (projectId, success, error) {
            return resource.GetDashboard({ action: 'GetDashboard', projectId: projectId }, success || noop, error || noop);
        };

        self.RequestExcel = function (success, error) {
            return resource.RequestExcel({ action: 'RequestExcel'}, success || noop, error || noop);
        };
        
    }]);
})(angular);